controllers/vm_controller.go

```
@@ -0,0 +1,140 @@
+	return vmControllerName
+}
+
+func (r *VmReconciler) Start(_ context.Context, mgr ctrl.Manager) error {
+	return r.setupController(mgr)
+}
+
+func newVmReconciler(mgr ctrl.Manager) (*VmReconciler, error) {
+	logger := ctrl.Log.WithName("controllers").WithName("VirtualMachines")
+	reconciler := &VmReconciler{
+		client: mgr.GetClient(),
+		log:    logger,
+	}
+
+	return reconciler, nil
+}
+
+func (r *VmReconciler) setupController(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(vmControllerName).
+		For(&kubevirtv1.VirtualMachine{}).
+		Complete(r)
+}
+
+func (r *VmReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	r.log.Info("Starting vm reconciliation...", "request", req.String())
+	vm := kubevirtv1.VirtualMachine{}
+	if err := r.client.Get(ctx, req.NamespacedName, &vm); err != nil {
+		if errors.IsNotFound(err) {
+			// VM was deleted, so we can ignore it
+			r.log.Info("VM not found", "vm", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		r.log.Error(err, "Error getting VM", "vm", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	if err := r.setVmVolumesMetrics(&vm); err != nil {
+		r.log.Error(err, "Error setting vm volumes metrics", "vm", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	r.log.Info("Finished vm reconciliation", "request", req.String())
+	return ctrl.Result{}, err
+}
+
+func (r *VmReconciler) setVmVolumesMetrics(vm *kubevirtv1.VirtualMachine) error {
+	var result error
+
+	if vm.Spec.Template.Spec.Volumes == nil {
+		return nil
+	}
+
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		volumeName := ""
+		if volume.DataVolume != nil {
+			volumeName = volume.DataVolume.Name
+		} else if volume.PersistentVolumeClaim != nil {
+			volumeName = volume.PersistentVolumeClaim.ClaimName
+		} else {
+			continue
+		}
+
+		pvc, err := r.getPVC(vm, volumeName)
+		if err != nil {
+			result = multierror.Append(result, err)
```

I think it would be better to return the error and end the loop early, because if this function returns an error, the reconciliation will be triggered again.

---

updated